applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_solver_vmsmonolithic.py

```diff
@@ -144,133 +133,160 @@ def PrepareModelPart(self):
+        if hasattr(self, "_turbulence_model_solver"):
+            self._turbulence_model_solver.SetCommunicator(self.get_epetra_communicator())
+
+        # Construct and set the solution strategy
+        solution_strategy = self.GetSolutionStrategy()
+        solution_strategy.SetEchoLevel(self.settings["echo_level"].GetInt())
+
+        # Initialize the solution strategy
+        if not self.is_restarted():
+            # Set the formulation instance process info variables
+            self.formulation.SetProcessInfo(self.GetComputingModelPart())
+            # Initialize the solution strategy
+            solution_strategy.Initialize()
+        else:
+            # This try is required in case SetInitializePerformedFlag is not a member of the strategy
+            try:
+                solution_strategy.SetInitializePerformedFlag(True)
+            except AttributeError:
+                pass
+
+        # If there is turbulence modelling, set the new solution strategy as parent strategy
         if hasattr(self, "_turbulence_model_solver"):
-            self._turbulence_model_solver.SetCommunicator(self.EpetraCommunicator)
+            self._turbulence_model_solver.SetParentSolvingStrategy(solution_strategy)
 
-        ## Get the computing model part
-        self.computing_model_part = self.GetComputingModelPart()
+        KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Solver initialization finished.")
 
-        ## If needed, create the estimate time step utility
-        if (self.settings["time_stepping"]["automatic_time_step"].GetBool()):
-            self.EstimateDeltaTimeUtility = self._GetAutomaticTimeSteppingUtility()
+    def Finalize(self):
+        self.GetSolutionStrategy().Clear()
 
-        ## Creating the Trilinos convergence criteria
-        if (self.settings["time_scheme"].GetString() == "bossak"):
-            self.conv_criteria = KratosTrilinos.TrilinosUPCriteria(self.settings["relative_velocity_tolerance"].GetDouble(),
-                                                                   self.settings["absolute_velocity_tolerance"].GetDouble(),
-                                                                   self.settings["relative_pressure_tolerance"].GetDouble(),
-                                                                   self.settings["absolute_pressure_tolerance"].GetDouble())
-        elif (self.settings["time_scheme"].GetString() == "steady"):
-            self.conv_criteria = KratosTrilinos.TrilinosResidualCriteria(self.settings["relative_velocity_tolerance"].GetDouble(),
-                                                                             self.settings["absolute_velocity_tolerance"].GetDouble())
+        if hasattr(self, "_turbulence_model_solver"):
+            self._turbulence_model_solver.Finalize()
 
-        (self.conv_criteria).SetEchoLevel(self.settings["echo_level"].GetInt())
+    def get_epetra_communicator(self):
```

Comment:
this is copy pasted in all trilinos solvers
=> Better refactor to a separate file (e.g. see here:https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/auxiliar_methods_adaptative_solvers.py)

Reply:
OK. Makes sense for me.

Reply:
However, how to make the Epetra communicator a member of this class if we implement the getter in an external file?